File: backend/src/lib/QuadraticCurveFitter.ts
import QuantityPricing from "./product/QuantityPricing";

// Quadratic Curve Fitter
export default class QuadraticCurveFitter {
  a: number;
  b: number;
  c: number;

  constructor() {
    this.a = 0;
    this.b = 0;
    this.c = 0;
  }

  fit(points: QuantityPricing[]): void {
    if (!points || points.length === 0) {
      throw new Error("Points list cannot be null or empty.");
    }

    const n = points.length;
    let sumX = 0,
      sumX2 = 0,
      sumX3 = 0,
      sumX4 = 0;
    let sumY = 0,
      sumXY = 0,
      sumX2Y = 0;

    points.forEach((p) => {
      sumX += p.quantity;
      sumX2 += p.quantity ** 2;
      sumX3 += p.quantity ** 3;
      sumX4 += p.quantity ** 4;
      sumY += p.price;
      sumXY += p.quantity * p.price;
      sumX2Y += p.quantity ** 2 * p.price;
    });

    const matrix = [
      [n, sumX, sumX2],
      [sumX, sumX2, sumX3],
      [sumX2, sumX3, sumX4],
    ];

    const vector = [sumY, sumXY, sumX2Y];

    const [c, b, a] = this.solve(matrix, vector);

    this.a = a;
    this.b = b;
    this.c = c;
  }

  predict(quantity: number): number {
    return this.a * quantity ** 2 + this.b * quantity + this.c;
  }

  solve(matrix: number[][], vector: number[]): number[] {
    const n = vector.length;
    const result = new Array(n).fill(0);

    for (let i = 0; i < n; i++) {
      for (let j = i + 1; j < n; j++) {
        const ratio = matrix[j][i] / matrix[i][i];
        for (let k = i; k < n; k++) {
          matrix[j][k] -= ratio * matrix[i][k];
        }
        vector[j] -= ratio * vector[i];
      }
    }

    for (let i = n - 1; i >= 0; i--) {
      result[i] = vector[i];
      for (let j = i + 1; j < n; j++) {
        result[i] -= matrix[i][j] * result[j];
      }
      result[i] /= matrix[i][i];
    }

    return result;
  }
}

File: price-engine/app/lib/quadratic_curve_fitter.py
from typing import List 

from .product.quantity_pricing import QuantityPricing

class QuadraticCurveFitter:
    def __init__(self):
        self.a = 0.0
        self.b = 0.0
        self.c = 0.0

    def fit(self, points: List[QuantityPricing]) -> None:
        if not points or len(points) == 0:
            raise ValueError("Points list cannot be null or empty.")

        if len(points) == 2:
            p1, p2 = points
            b = (p2.price - p1.price) / (p2.quantity - p1.quantity)
            c = p1.price - b * p1.quantity

            self.a = 0.0  # No quadratic term
            self.b = b
            self.c = c
            return

        n = len(points)
        sum_x = sum_x2 = sum_x3 = sum_x4 = 0.0
        sum_y = sum_xy = sum_x2y = 0.0

        for p in points:
            x = p.quantity
            y = p.price
            sum_x += x
            sum_x2 += x ** 2
            sum_x3 += x ** 3
            sum_x4 += x ** 4
            sum_y += y
            sum_xy += x * y
            sum_x2y += x ** 2 * y

        matrix = [
            [n, sum_x, sum_x2],
            [sum_x, sum_x2, sum_x3],
            [sum_x2, sum_x3, sum_x4],
        ]

        vector = [sum_y, sum_xy, sum_x2y]

        c, b, a = self.solve(matrix, vector)

        self.a = a
        self.b = b
        self.c = c

    def predict(self, quantity: float) -> float:
        return self.a * quantity ** 2 + self.b * quantity + self.c

    def solve(self, matrix: List[List[float]], vector: List[float]) -> List[float]:
        n = len(vector)
        result = [0.0 for _ in range(n)]

        # Forward elimination
        for i in range(n):
            for j in range(i + 1, n):
                ratio = matrix[j][i] / matrix[i][i]
                for k in range(i, n):
                    matrix[j][k] -= ratio * matrix[i][k]
                vector[j] -= ratio * vector[i]

        # Back substitution
        for i in range(n - 1, -1, -1):
            result[i] = vector[i]
            for j in range(i + 1, n):
                result[i] -= matrix[i][j] * result[j]
            result[i] /= matrix[i][i]

        return result
